From 34e78f2517c4981e972539d2ae48d3a63a789e5a Mon Sep 17 00:00:00 2001
From: Stephen <stephen@vamrs.com>
Date: Mon, 11 Oct 2021 17:48:03 +0800
Subject: [PATCH 07/17] arm64: dts: add meson i2c and spi overlay support

Signed-off-by: Stephen <stephen@vamrs.com>
---
 arch/arm64/boot/dts/amlogic/overlay/Makefile  |  8 ++-
 .../dts/amlogic/overlay/README.meson-overlays | 66 ++++++++++++++++---
 .../dts/amlogic/overlay/meson-fixup.scr-cmd   | 14 ++++
 ...eson-g12a-i2c-ao-m0-gpioao-2-gpioao-3.dts} |  0
 ... meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7.dts} |  2 +-
 ...eson-g12a-i2c-ee-m1-gpiox-10-gpiox-11.dts} |  2 +-
 ...eson-g12a-i2c-ee-m3-gpioa-14-gpioa-15.dts} |  0
 .../amlogic/overlay/meson-g12a-spi-spidev.dts | 38 +++++++++++
 8 files changed, 115 insertions(+), 15 deletions(-)
 rename arch/arm64/boot/dts/amlogic/overlay/{meson-g12a-i2c-ao-m0.dts => meson-g12a-i2c-ao-m0-gpioao-2-gpioao-3.dts} (100%)
 copy arch/arm64/boot/dts/amlogic/overlay/{meson-g12a-i2c-ee-m1.dts => meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7.dts} (77%)
 rename arch/arm64/boot/dts/amlogic/overlay/{meson-g12a-i2c-ee-m1.dts => meson-g12a-i2c-ee-m1-gpiox-10-gpiox-11.dts} (77%)
 rename arch/arm64/boot/dts/amlogic/overlay/{meson-g12a-i2c-ee-m3.dts => meson-g12a-i2c-ee-m3-gpioa-14-gpioa-15.dts} (100%)
 create mode 100644 arch/arm64/boot/dts/amlogic/overlay/meson-g12a-spi-spidev.dts

diff --git a/arch/arm64/boot/dts/amlogic/overlay/Makefile b/arch/arm64/boot/dts/amlogic/overlay/Makefile
index 5b862d4ee4c3..ddf7057abb15 100644
--- a/arch/arm64/boot/dts/amlogic/overlay/Makefile
+++ b/arch/arm64/boot/dts/amlogic/overlay/Makefile
@@ -1,10 +1,12 @@
 # SPDX-License-Identifier: GPL-2.0
 dtbo-$(CONFIG_ARCH_MESON) += \
-	meson-g12a-i2c-ee-m1.dtbo \
-	meson-g12a-i2c-ee-m3.dtbo \
-	meson-g12a-i2c-ao-m0.dtbo \
+	meson-g12a-i2c-ao-m0-gpioao-2-gpioao-3.dtbo \
+	meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7.dtbo \
+	meson-g12a-i2c-ee-m1-gpiox-10-gpiox-11.dtbo \
+	meson-g12a-i2c-ee-m3-gpioa-14-gpioa-15.dtbo \
 	meson-g12a-pwmao-a-on-gpioao-11.dtbo \
 	meson-g12a-pwm-c-on-gpiox-8.dtbo \
+	meson-g12a-spi-spidev.dtbo \
 	meson-g12a-uart-ee-c.dtbo
 
 scr-$(CONFIG_ARCH_MESON) += \
diff --git a/arch/arm64/boot/dts/amlogic/overlay/README.meson-overlays b/arch/arm64/boot/dts/amlogic/overlay/README.meson-overlays
index 956d173b159b..a35bc552dc6f 100644
--- a/arch/arm64/boot/dts/amlogic/overlay/README.meson-overlays
+++ b/arch/arm64/boot/dts/amlogic/overlay/README.meson-overlays
@@ -16,7 +16,9 @@ Kernel provided DT overlay files are in
 
 /boot/uEnv.txt contents:
 
-overlays=meson-g12a-i2c-ee-m3 meson-g12a-i2c-ao-m0 meson-g12a-uart-ee-c
+overlays=meson-g12a-i2c-ao-m0-gpioao-2-gpioao-3 meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7 meson-g12a-i2c-ee-m3-gpioa-14-gpioa-15 meson-g12a-spi-spidev
+param_spidev_spi_bus=0
+param_spidev_max_freq=10000000
 
 Add names of overlays you want to activate to overlays= line in /boot/uEnv.txt, separated with spaces.
 Add required parameters with their values to /boot/uEnv.txt, one per line.
@@ -24,22 +26,34 @@ Add optional parameters with their values to /boot/uEnv.txt if you want to chang
 
 #### Meson Overlay details:
 
-- meson-g12a-i2c-ee-m1 meson-g12a-i2c-ee-m3 meson-g12a-i2c-ao-m0
+- meson-g12a-i2c-ao-m0-gpioao-2-gpioao-3 meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7 meson-g12a-i2c-ee-m1-gpiox-10-gpiox-11 meson-g12a-i2c-ee-m3-gpioa-14-gpioa-15
+- meson-g12a-pwmao-a-on-gpioao-11 meson-g12a-pwm-c-on-gpiox-8
+- meson-g12a-spi-spidev
 - meson-g12a-uart-ee-c
-- meson-g12a-pwm-c-on-gpiox-8 meson-g12a-pwmao-a-on-gpioao-11
 
-#### meson-g12a-i2c-ee-m1
+#### meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7
 
-Activates I2C EE M3
+Activates I2C EE M1
 
 Device name: /dev/i2c-1
 
 Pins (SCL, SDA): GPIOH_7, GPIOH_6
 
 Usage: add one line in /boot/uEnv.txt
-overlays=meson-g12a-i2c-ee-m1
+overlays=meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7
+
+#### meson-g12a-i2c-ee-m1-gpiox-10-gpiox-11
+
+Activates I2C EE M1
+
+Device name: /dev/i2c-1
+
+Pins (SCL, SDA): GPIOX_11, GPIOX_10
+
+Usage: add one line in /boot/uEnv.txt
+overlays=meson-g12a-i2c-ee-m1-gpiox-10-gpiox-11
 
-#### meson-g12a-i2c-ee-m3
+#### meson-g12a-i2c-ee-m3-gpioa-14-gpioa-15
 
 Activates I2C EE M3
 
@@ -48,9 +62,9 @@ Device name: /dev/i2c-3
 Pins (SCL, SDA): GPIOA_15, GPIOA_14
 
 Usage: add one line in /boot/uEnv.txt
-overlays=meson-g12a-i2c-ee-m3
+overlays=meson-g12a-i2c-ee-m3-gpioa-14-gpioa-15
 
-#### meson-g12a-i2c-ao-m0
+#### meson-g12a-i2c-ao-m0-gpioao-2-gpioao-3
 
 Activates I2C AO M0
 
@@ -59,7 +73,7 @@ Device name: /dev/i2c-4
 Pins (SCL, SDA): GPIOAO_2, GPIOAO_3
 
 Usage: add one line in /boot/uEnv.txt
-overlays=meson-g12a-i2c-ao-m0
+overlays=meson-g12a-i2c-ao-m0-gpioao-2-gpioao-3
 
 #### meson-g12a-uart-ee-c
 
@@ -82,3 +96,35 @@ Activates PWMAO_A
 
 Pins: GPIOAO_11
 
+#### meson-g12a-spi-spidev
+
+Activates SPIdev device node (/dev/spidevX.Y) for userspace SPI access,
+where X is the bus number and Y is the CS number
+
+SPI A bus
+    pins (SS0, SCLK, MOSI, MISO): GPIOX_10, GPIOX_11, GPIOX_8, GPIOX_9
+    device: /dev/spidev0.0
+
+SPI B bus
+    pins (SS0, SCLK, MOSI, MISO): GPIOH_6, GPIOH_7, GPIOH_4, GPIOH_5
+    device: /dev/spidev1.0
+
+Parameters:
+
+param_spidev_spi_bus (int)
+    SPI bus to activate SPIdev support on
+    Required
+    Supported values: 0, 1
+
+param_spidev_spi_cs (int)
+    SPI chip select number
+    Optional
+    Default: 0
+    Supported values: 0
+
+param_spidev_max_freq (int)
+    Maximum SPIdev frequency
+    Optional
+    Default: 10000000
+    Range: 3000 - 100000000
+
diff --git a/arch/arm64/boot/dts/amlogic/overlay/meson-fixup.scr-cmd b/arch/arm64/boot/dts/amlogic/overlay/meson-fixup.scr-cmd
index d4c39e20a3a2..067b61b80de1 100644
--- a/arch/arm64/boot/dts/amlogic/overlay/meson-fixup.scr-cmd
+++ b/arch/arm64/boot/dts/amlogic/overlay/meson-fixup.scr-cmd
@@ -2,3 +2,17 @@
 # implements (or rather substitutes) overlay arguments functionality
 # using u-boot scripting, environment variables and "fdt" command
 
+if test -n "${param_spidev_spi_bus}"; then
+	test "${param_spidev_spi_bus}" = "0" && setenv tmp_spi_path "soc/bus@ffd00000/spi@13000"
+	test "${param_spidev_spi_bus}" = "1" && setenv tmp_spi_path "soc/bus@ffd00000/spi@15000"
+	test "${param_spidev_spi_bus}" = "2" && setenv tmp_spi_path "soc/bus@ffd00000/spi@14000"
+	fdt set /${tmp_spi_path} status "okay"
+	fdt set /${tmp_spi_path}/spidev status "okay"
+	if test -n "${param_spidev_max_freq}"; then
+		fdt set /${tmp_spi_path}/spidev@0 spi-max-frequency "<${param_spidev_max_freq}>"
+	fi
+	if test "${param_spidev_spi_cs}" = "1"; then
+		fdt set /${tmp_spi_path}/spidev reg "<1>";
+	fi
+fi
+
diff --git a/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ao-m0.dts b/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ao-m0-gpioao-2-gpioao-3.dts
similarity index 100%
rename from arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ao-m0.dts
rename to arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ao-m0-gpioao-2-gpioao-3.dts
diff --git a/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1.dts b/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7.dts
similarity index 77%
copy from arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1.dts
copy to arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7.dts
index 0b7915d122fb..a68284bf61b8 100644
--- a/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1.dts
+++ b/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1-gpioh-6-gpioh-7.dts
@@ -8,7 +8,7 @@ fragment@0 {
 		target = <&i2c1>;
 		__overlay__ {
 			status = "okay";
-			pinctrl-0 = <&i2c1_sck_x_pins &i2c1_sda_h6_pins>;
+			pinctrl-0 = <&i2c1_sda_h6_pins &i2c1_sck_h7_pins>;
 			pinctrl-names = "default";
 		};
 	};
diff --git a/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1.dts b/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1-gpiox-10-gpiox-11.dts
similarity index 77%
rename from arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1.dts
rename to arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1-gpiox-10-gpiox-11.dts
index 0b7915d122fb..eb9c402fb633 100644
--- a/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1.dts
+++ b/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m1-gpiox-10-gpiox-11.dts
@@ -8,7 +8,7 @@ fragment@0 {
 		target = <&i2c1>;
 		__overlay__ {
 			status = "okay";
-			pinctrl-0 = <&i2c1_sck_x_pins &i2c1_sda_h6_pins>;
+			pinctrl-0 = <&i2c1_sda_x_pins &i2c1_sck_x_pins>;
 			pinctrl-names = "default";
 		};
 	};
diff --git a/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m3.dts b/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m3-gpioa-14-gpioa-15.dts
similarity index 100%
rename from arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m3.dts
rename to arch/arm64/boot/dts/amlogic/overlay/meson-g12a-i2c-ee-m3-gpioa-14-gpioa-15.dts
diff --git a/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-spi-spidev.dts b/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-spi-spidev.dts
new file mode 100644
index 000000000000..abb23385a407
--- /dev/null
+++ b/arch/arm64/boot/dts/amlogic/overlay/meson-g12a-spi-spidev.dts
@@ -0,0 +1,38 @@
+/dts-v1/;
+/plugin/;
+
+/ {
+	compatible = "radxa,zero", "amlogic,g12a";
+
+	fragment@0 {
+		target = <&spicc0>;
+		__overlay__ {
+			pinctrl-0 = <&spicc0_x_pins &spicc0_ss0_x_pins>;
+			pinctrl-names = "default";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			spidev@0 {
+				compatible = "spidev";
+				status = "disabled";
+				reg = <0>;
+				spi-max-frequency = <10000000>;
+			};
+		};
+	};
+
+	fragment@1 {
+		target = <&spicc1>;
+		__overlay__ {
+			pinctrl-0 = <&spicc1_pins &spicc1_ss0_pins>;
+			pinctrl-names = "default";
+			#address-cells = <1>;
+			#size-cells = <0>;
+			spidev@0 {
+				compatible = "spidev";
+				status = "disabled";
+				reg = <0>;
+				spi-max-frequency = <10000000>;
+			};
+		};
+	};
+};
-- 
2.33.0

